### ISSUE TITLE
Question: is it even possible to have multiple "StdCF" crypto filters?

### ISSUE THREAD

**Issue by test-PDF** [pdf-association/pdf-issues#89]
In the 7.6.4.1 "Standard security handler, General" there is this sentence:

> If a security handler of revision 4 or 5 is specified, the standard security handler shall support crypt filters (see 7.6.6, "Crypt filters"). The support shall be limited to the Identity crypt filter (see "Table 26: Standard crypt filter names") and crypt filters named StdCF whose dictionaries <...>

So, in fact due to nature of "password-based encryption" is is not possible to have multiple variants of passwords (there are only values for "U" etc. in "encryption dictionary", not in "crypt filter dictionary"). But at the same time standard said that there are "crypt **FILTERS** named StdCF". Can anyone clarify me how to use "multiple StdCFs" here?

**Comment by MatthiasValvekens** [comment]
Multiple password-based (standard) crypt filters is not something that makes sense in PDF right now, that's right. However, you can have PDFs where embedded files are encrypted but the rest of the document is not, or PDFs where all strings/streams are encrypted but metadata is not. This is achieved by combining the **Identity** crypt filter with a regular **StdCF** one (again, assuming that we're using the standard security handler).

In a document protected by a public key security handler (as opposed to the standard one), it's not so hard to imagine a situation where different embedded files are encrypted using different keys, and hence different crypt filters, for example. The standard allows for that kind of thing (that's why **Recipients** is a crypt filter property, not a security handler property for public key security handlers). That being said, I'm sure that very few viewers, if any, would support it. You could go even more wild with custom security handlers & crypt filters if your use case requires it.

Long story short, the crypt filter mechanism is very flexible, but for interoperability reasons, only a very specific set of configurations is actually practical to use for public workflows.

@petervwyatt I think this should get the "question" label, but I can't assign it :)

**Comment by test-PDF** [comment]
@MatthiasValvekens Sure, thank you for the clarification. Yes, I do understand that there is a difference between "password-based" and "certificate-based" encryptions. And I do understand that could be a situation when for some internal objects "Identity" crypt filter would be using. But my question: is it even possible to have multiple StdCF crypto filters inside CF dictionary? The standard said IT IS POSSIBLE, but I cannot imagine how to use it.

**Comment by MatthiasValvekens** [comment]
Right, I think I might have made that edit while you were typing up your comment, so let me copy it here:

> Is your question about why "filters" is plural in "crypt filters named StdCF"? I think that's just a generic plural; at least in my
> view, it doesn't imply the potential presence of multiple StdCF filters in a single document---that would be impossible anyway,
> since crypt filters are in a document-wide namespace.

Does that answer your question, or is there something I'm misunderstanding here?

**Comment by test-PDF** [comment]
@MatthiasValvekens Yes, it is close to answer, at least now I know your opinion :) But maybe we need to change PDF standard in this sentence? It is really unclear.

**Comment by test-PDF** [comment]
I do think would be enough to remove just one "...s" and have something like

> and crypt filter named StdCF whose dictionaries <...>

**Comment by MatthiasValvekens** [comment]
Right, I personally don't see the ambiguity, but I'm also not necessarily opposed to making it singular instead :)

**Comment by test-PDF** [comment]
BTW, can anyone answer me related question: why for any good reason to have a "dedicated crypto filter name" here? I mean why strictly said "only StdCF name allowed"? The name is using only once: in StrF/StmF values inside "encryption dictionary". So, if any other name would be using instead of StdCF nothing bad happens.

**Comment by lrosenthol** [comment]
It's historical @test-PDF - based on the original hope that multiple and custom crypt filters would be supported.  However, that never happened and so we only have a single defined name.

Perhaps in the future we will have more...

**Comment by petervwyatt** [comment]
The full para in question is:

> If a security handler of revision 4 or 5 is specified, the standard security handler shall support crypt filters (see 7.6.6, "Crypt filters"). The support shall be limited to the Identity crypt filter (see "Table 26 — Standard crypt filter names") and crypt filters named StdCF whose dictionaries contain an AuthEvent value of DocOpen. For revision 4, the filter CFM value shall be V2 (RC4) or AESV2 (AES-128). For revision 6, the filter CFM value shall be AESV3 (AES-256). Public-Key security handlers in this case shall use crypt filters named DefaultCryptFilter when all document content is encrypted, and shall use crypt filters named DefEmbeddedFile when file attachments only are encrypted in place of StdCF name. This nomenclature shall not be used as an indicator of the type of the security handler or encryption. Use of security handler revisions 1, 2, 3, 4 and 5 is deprecated in PDF 2.0.

so I can see how non-native English speakers might get confused with the unnecessary plurals in a few places. Here's a suggested fix shown in italics (note that the natural formatting of bold key names and italic key values from ISO 32K-2 has also been removed so the changes are more obvious):

> If a security handler of revision 4 or 5 is specified, the standard security handler shall support crypt filters (see 7.6.6, "Crypt filters"). The support shall be limited to the Identity crypt filter (see "Table 26 — Standard crypt filter names") and _a crypt filter_ named StdCF whose dictionaries contain an AuthEvent value of DocOpen. For revision 4, the filter CFM value shall be V2 (RC4) or AESV2 (AES-128). For revision 6, the filter CFM value shall be AESV3 (AES-256). Public-Key security handlers in this case shall use _a crypt filter_ named DefaultCryptFilter when all document content is encrypted, and shall use _a crypt filter_ named DefEmbeddedFile when file attachments only are encrypted in place of StdCF name. This nomenclature shall not be used as an indicator of the type of the security handler or encryption. Use of security handler revisions 1, 2, 3, 4 and 5 is deprecated in PDF 2.0.

**Comment by test-PDF** [comment]
@petervwyatt Thank you, looking forward to have the proposal in PDF standard.

**Comment by petervwyatt** [comment]
PDF TWG agree
